Add new logic to process credits information into the JS definition

sanderfrenken opened the pull request:
This PR adds a new way to handle the licenses, partly related to:

[issue: Add a License Filter](https://github.com/sanderfrenken/Universal-LPC-Spritesheet-Character-Generator/issues/210)

In this PR, a script is added that traverses all json files.
From each file, it checks for all the variants (the actual image) if there exists an entry in the credits.csv.
Moreover, if there is a parent node  in the credits.csv instead of the actual variant, it will use that parent information. If both are there, the logic prefers to only use the parent node. This is done to keep the license information in the json file as condensed as possible.

For instance:

inside `arms_armour.json` you will now find:

```
"credits": [
    {
      "file": "arms/armour/plate/male",
      "notes": "original by wulax, recolors by bigbeargames; adapted to v3 bases and further recolors by bluecarrot16",
      "authors": [
        "Michael Whitlock (bigbeargames)",
        "Matthew Krohn (makrohn)",
        "Johannes Sjölund (wulax)"
      ],
      "licenses": [
        "OGA-BY 3.0",
        "CC-BY-SA 3.0",
        "GPL 3.0"
      ],
      "urls": [
        "https://opengameart.org/content/lpc-medieval-fantasy-character-sprites",
        "http://opengameart.org/content/lpc-clothing-updates"
      ]
    },
    {
      "file": "arms/armour/plate/female",
      "notes": "original by wulax, adapted to female by makrohn, recolors by bigbeargames; adapted to v3 bases and further recolors by bluecarrot16",
      "authors": [
        "Michael Whitlock (bigbeargames)",
        "Matthew Krohn (makrohn)",
        "Johannes Sjölund (wulax)"
      ],
      "licenses": [
        "OGA-BY 3.0",
        "CC-BY-SA 3.0",
        "GPL 3.0"
      ],
      "urls": [
        "https://opengameart.org/content/lpc-medieval-fantasy-character-sprites",
        "https://opengameart.org/content/lpc-combat-armor-for-women",
        "http://opengameart.org/content/lpc-clothing-updates"
      ]
    }
  ]
```

Instead of having an entry for each variant. This can be done, as in the credits.csv these assets are represented as:

```
arms/armour/plate/female,"original by wulax, adapted to female by makrohn, recolors by bigbeargames; adapted to v3 bases and further recolors by bluecarrot16 ","Michael Whitlock (bigbeargames), Matthew Krohn (makrohn), Johannes Sjölund (wulax)","OGA-BY 3.0, CC-BY-SA 3.0, GPL 3.0",https://opengameart.org/content/lpc-medieval-fantasy-character-sprites,https://opengameart.org/content/lpc-combat-armor-for-women,http://opengameart.org/content/lpc-clothing-updates,,,,,,,,,,,,,

arms/armour/plate/male,"original by wulax, recolors by bigbeargames; adapted to v3 bases and further recolors by bluecarrot16 ","Michael Whitlock (bigbeargames), Matthew Krohn (makrohn), Johannes Sjölund (wulax)","OGA-BY 3.0, CC-BY-SA 3.0, GPL 3.0",https://opengameart.org/content/lpc-medieval-fantasy-character-sprites,http://opengameart.org/content/lpc-clothing-updates,,,,,,,,,,,,,,
```

Which indicates that all the variants use the same licensing information as these parents.

At the same time, you can see that for instance in the hair json files there is a huge amount of repetition. That is because in the credits.csv there are no parents for these entries added. 

I would like to clean that up as much as possible, and then be able to have clean json files with the minimum required licensing info. 

From that moment, I can also adjust the HTML content to actually contain the license information per file. 

Then we can remove the parsing csv logic from the `chargen.js`. and we can remove the `credits.csv` and related logic, and rely on the JSON files instead for the license information.

Commit:
Add new logic to process credits information into the JS definition

jrconway3 commented:
Okay... Okay, lol... I was worried you went and manually updated every file, but looks like you create a script to generate it. That's cool. I'm reinstalling XAMPP now so I can test this locally.


That being said, we shouldn't get rid of the credits.csv file. Perhaps what we can do is sync between the two, but rely on the data in the sheets. The reason being is that the credits.csv file is linked in various places like on OGA releases, so we should still keep it in some form at least for legacy reasons.

jrconway3 commented:
"At the same time, you can see that for instance in the hair json files there is a huge amount of repetition. That is because in the credits.csv there are no parents for these entries added."

Yeah, that's because I've only been adding the parents as I fill out the assets. Hair is something I have barely touched because the hair, realistically speaking, needs animations to make it look better on run/jump/etc. Short hair is fine, though, maybe soon I can go through and fill out all the really short hair, but longer hair needs animation to make it look better.


"If both are there, the logic prefers to only use the parent node."

This I am concerned with. I think we need to simply clean out the credits.csv to remove all duplicates first. I can go through and clean out the whole CSV if you like.

We need to handle deepest elements as a priority because some variants may have been made by different people. It should fall back to the parent only if a more specific element doesn't exist.

About the only thing that probably isn't very notable are recolors, unless the recolors are doing a heavy repalette (color reduction for example). If its actually making significant changes instead of simply swapping colors, then that needs an override.

sanderfrenken commented:
> This I am concerned with. I think we need to simply clean out the credits.csv to remove all duplicates first. I can go through and clean out the whole CSV if you like.

Yes I can elaborate there: it only will use the parent if the children have the same licenses and authors. So it does check if indeed the more specific credit is not different from the parent. If it is different, it will add the specific license instead of the parent for that variant.

If you can have a look at the CSV, that would be nice. But I will do it otherwise coming Friday. So don't feel obligated to do it if it doesn't fit your schedule :)

> Perhaps what we can do is sync between the two, but rely on the data in the sheets. The reason being is that the credits.csv file is linked in various places like on OGA releases, so we should still keep it in some form at least for legacy reasons.

Yes you are right about the sheet references. Also there is a button to download the full sheet from the generator. So indeed we will leave it in. I will add the syncing mechanism when its finished to the generate_html script.

jrconway3 commented:
Finally got my localhost running again and I have my vhost working. No more certificate errors and no more ulpc.localhost/ulpc with /ulpc being a symlink to the actual directory...

I'll run some tests on this later tonight. :P

Commit:
* Merged Conflicts

Commit:
* Removed Dups for Simple Earring

Commit:
* Cleaned Full Credits List

Commit:
* Fixed Issues With Credits.csv not formatted correctly

jrconway3 commented:
Right now the credits doesn't seem to work for me on this branch. I'll try running the command soon. I just opened a PR into this branch with the credits list fix. I also added OGA-BY to a number of assets as I'm pretty sure they should be OGA-BY (this includes a number of assets by Nilla and Wulax that still didn't have OGA-BY added yet). There's probably some more I missed as well.

Commit:
Update credits

Commit:
Implement new credit creation from JSON info

sanderfrenken commented:
Thanks for the updated CSV! It now looks very clean. If you run the command 

`node scripts/node scripts/generate_html.js`

You will now see warnings for a couple of files left that don't hold (complete) license information. For instance, for the muscular bodies there are no licenses defined in the credits.csv, and neither for the legs/pants/pregnant entries.

I updated the generator logic. Indeed the crediting was not working before, I only made the script at first. But now its almost done.

- All information for the credits is taken from the JSON files. So I could remove a lot of CSV related logic. IMO the code is more readable now and easier to maintain.

- The credits are created for now from the CSV file. But when all data is present in the json files, I prefer to go the other way around: Generate the CSV from the JSON's. Then we can link still to this CSV in the repo for downloading all credits if needed. The CSV will receive an entry for each individual file in the generator. It will autogenerate when running the generate_html script. The generate_html will also throw an error when there are no credits for a certain file (now it just logs it). That way, we always have proper credits in the generator.

- I Still need to implement the export sheet credits to CSV option. Its now a placeholder in the source: 

```
function sheetCreditsToCSV() {
    // TODO
}
```

- I will also clean up the python files related to creating the current credits csv.
- I will also clean up the JS script to generate the license info from the CSV as that is a one time action
- Now I can also add a filter for a specific license. That should be rather easy now that the license information is inside the JSON. I hope...

Commit:
remove parse-csv.js from index import

sanderfrenken commented:
Oh and btw with this change you also don't need a server anymore to run the project without adjusting the code. A nice side effect :)

Commit:
Update credit labels and explanation

Commit:
Pick the right credit when there are multiple entries for a variant

Commit:
Minor fixes

Commit:
Credit fixes

jrconway3 commented:
Cool, I'll run some tests on it tonight.

"You will now see warnings for a couple of files left that don't hold (complete) license information. For instance, for the muscular bodies there are no licenses defined in the credits.csv, and neither for the legs/pants/pregnant entries."

Weird, I didn't see this issue. I'm not sure why that would be the case. When running the CSV normally in the master branch, everything worked fine there.


I noticed you added kite shield. I tried making sure the parents were configured correctly and I didn't see kite as a parent.

sanderfrenken commented:
Yes on master the issue is also present:

https://sanderfrenken.github.io/Universal-LPC-Spritesheet-Character-Generator/#?body=Body_color_light&head=Human_male_light&sex=muscular

You can see the license field is empty (the other credit fields are populated properly though). 

Indeed I added the kite shields to their own folder. Makes the crediting easier plus it's better organized this way imo.

jrconway3 commented:
Its weird its broken like that when the console isn't reporting that the credit is missing...

jrconway3 commented:
Yeah so the license on muscular is literally missing but everything else is present. I was certain it was there before.

EDIT: OGA-BY is missing from Male Body? I was certain it was there before as well.

sanderfrenken commented:
> Its weird its broken like that when the console isn't reporting that the credit is missing...

Hmm not really. The logic that warns here is not really failsafe:

https://github.com/sanderfrenken/Universal-LPC-Spritesheet-Character-Generator/blob/e87766c66ff1bed22b1c6138ffe5d2e1b6668335/sources/chargen.js#L68

It only checks if there is any credits entry, and it only checks it for the first layer of the asset.
I also removed it in this branch btw.

sanderfrenken commented:
> Yeah so the license on muscular is literally missing but everything else is present. I was certain it was there before.
> 
> EDIT: OGA-BY is missing from Male Body? I was certain it was there before as well.

On master already as well. It is there for the head though, maybe you confuse them?

https://sanderfrenken.github.io/Universal-LPC-Spritesheet-Character-Generator/#?body=Body_color_light&head=Human_male_light

Or it got lost somewhere along the way. I think that is another argument to switch to having these credits in the json files, its much easier to review them.

jrconway3 commented:
Right, I get what you're saying. I looked it up immediately after I left that comment because I was confused as to why the entry was missing and that's when I realized it was literally just that one column missing (I know the failsafe only works if the entire row is missing).

However, I was certain the license _was_ set before... At least for Muscular that is. I remember looking it up and being like "oh its disappointing that Muscular isn't in OGA-BY".

jrconway3 commented:
Yes, nice. Nice. I see now that the credits are showing up properly and they're nice and clean. I'll run a test to generate adjustments and see how that looks.

Here for example:
```
body/bodies/male/light.png
	- Note: see details at https://opengameart.org/content/lpc-character-bases; 'Thick' Male Revised Run/Climb by JaidynReiman (based on ElizaWy's LPC Revised)
	- Licenses:
		- CC-BY-SA 3.0
		- GPL 3.0
	- Authors:
		- bluecarrot16
		- Benjamin K. Smith (BenCreating)
		- Evert
		- Eliza Wyatt (ElizaWy)
		- TheraHedwig
		- MuffinElZangano
		- Durrani
		- Johannes Sj?lund (wulax)
		- Stephen Challener (Redshrike)
	- Links:
		- https://opengameart.org/content/liberated-pixel-cup-lpc-base-assets-sprites-map-tiles
		- https://opengameart.org/content/lpc-medieval-fantasy-character-sprites
		- https://opengameart.org/content/lpc-male-jumping-animation-by-durrani
		- https://opengameart.org/content/lpc-runcycle-and-diagonal-walkcycle
		- https://opengameart.org/content/lpc-revised-character-basics
		- https://opengameart.org/content/lpc-be-seated
		- https://opengameart.org/content/lpc-runcycle-for-male-muscular-and-pregnant-character-bases-with-modular-heads
		- https://opengameart.org/content/lpc-jump-expanded
		- https://opengameart.org/content/lpc-character-bases

head/heads/human/male/light.png
	- Note: original head by Redshrike, tweaks by BenCreating, modular version by bluecarrot16
	- Licenses:
		- OGA-BY 3.0
		- CC-BY-SA 3.0
		- GPL 3.0
	- Authors:
		- bluecarrot16
		- Benjamin K. Smith (BenCreating)
		- Stephen Challener (Redshrike)
	- Links:
		- https://opengameart.org/content/liberated-pixel-cup-lpc-base-assets-sprites-map-tiles
		- https://opengameart.org/content/lpc-character-bases
```

I think this is really nice and clean. I quite like this structure. That said, it will make the credits quite long, but the credits will be long anyway, so this probably isn't a huge concern.

About the only other argument that could be made is listing off the authors separately and showing what assets they contributed to. But we can just have users decide how they want to format the credits and simply display it this way based on each asset.

jrconway3 commented:
Ran some tests on the generation and its looking really good, too. I think the only thing that's left now is adding the filter and making it generate when running generate_html.js.

sanderfrenken commented:
> Ran some tests on the generation and its looking really good, too. I think the only thing that's left now is adding the filter and making it generate when running generate_html.js.

Hee there! Yes its almost done, thanks for having a look :)

I also still want to clean some old credits related stuff if possible. I am thinking we could actually ditch the credits.csv and therewith the need to generate the credits into the jsons, as we could just work directly in the json's instead from now on.

Not sure though if that is a good idea, what is your opinion?

jrconway3 commented:
I have no problem with working directly from the json files, but like I said before: we should still support the credits.csv for backwards compatibility. Some places such as on OGA do link directly to it. So if you still implement the reverse-generation of the credits.csv from the json files, I think we should be good there.

But of course we also need to alert anyone else still using it that the change has been made.

sanderfrenken commented:
> So if you still implement the reverse-generation of the credits.csv from the json files, I think we should be good there.

That I will do:) 

So things left:

- generate credits.csv from jsons
- fix missing credits
- cleanup remaining csv related items
- add filter
- implement the export sheet credits to CSV option

Will finalize this somewhere in this week I hope :)

jrconway3 commented:
I'm dropping another batch of updates soon-ish which I think will be my final update for the near future. (Like a couple weeks probably). Once I do that, I'll try and also merge that branch into the licensing branch to fix the conflicts in the credits.

Commit:
* Merged Master and Fixed Conflicts

Commit:
* Added Missing Tool, Body, Visor, and Pregnant Pants Credits

Commit:
fix rebasing

Commit:
Fix export to CSV

sanderfrenken commented:
OK I rebased, which was a bit annoying, and fixed the export to CSV functionality.

So in essence, everything is now working nicely more or less as is. I like to open a new PR though to clean up the credits.csv and instead create it from the json information. It will keep the PR a bit cleaner and easier to review, and then at least I don't need to rebase again tomorrow etc.

So in subsequent PR's:

1. license filter
2. create csv from json files instead (so from that point, the JSON's will be the point of truth wrt the crediting)
3. disable non-credited files from being part of the generator (at the moment `belt_waistband.json` and `hair_long_center_part.json`)

I will also add the license filtering separately. Can you check this PR if it's all ok? I think it is so I would like to merge and continue with the remaining steps :)

jrconway3 commented:
Sure I'll take a look tonight.

jrconway3 commented:
So I can't see anything wrong with this. I did try out the CSV export as well; I assumed you were referring to converting the sheets into CREDITS.csv, but I guess that's still not here yet, rather this was about the "Export Credits (CSV)" button, and that works fine. I also tested adding new entries (namely the missing waistband and long_center_part hair) and those work fine, too.

Commit:
add final missing credits
